MILESTONES.md

```diff
@@ -0,0 +1,152 @@
+      can thus also include the JetBrains state version/serial to the hats file
+        - This isn't 100% guaranteed to be race-proof if multiple JetBrains
+          changes come in quickly, but hopefully it's close enough
+
+### Problems:
+
+1. Cursorless commands need to be blocking, and when you use VS Code, they are
+   because of the command server (the action does not return until the command
+   is fully applied.).
+    - However, with the sidecar, the _transmission_ of that cursorless change
+      back to JetBrains would be asynchronous.
+    - This means that something like a cursorless command followed by "air"
+      would have those two commands race.
+        - So you won't really be able to chain cursorless commands alongside
+          other Talon actions unless they do not logically affect the same part
+          of the code.
+        - This is fine for M0.1 because the goal is just to get something
+          working, but will to revisit this for M0.2.
+
+2. The VS Code command server requires a keystroke to be triggered for it to "
+   pick up" the command and respond to it. This requires that VS Code be
+   focused, which obviously wouldn't be the case if the user intends it to be a
+   background sidecar.
+    - On macOS we can work around this because there's an API to send keystrokes
+      to non focused applications, but not on other platforms.
+    - Solution: if you don't care about using VS Code outside of cursorless
+      everywhere / are less paranoid about security, we could fork the command
+      server to be an HTTP server which would not require this keystroke, just
+      for this prototype.
+        - Why wasn't it an HTTP server originally? Security concerns from a
+          user. However it shouldn't be a huge deal -- the existing JetBrains
+          plugin uses a HTTP server with a nonce file.
+        - The other reason we have the keystroke is to provide a synchronization
+          barrier if the user is using VS Code as their primary editor.
+            - Many Talon commands insert by mimicking keystrokes, so the
+              shortcut helps ensure that cursorless commands, which would
+              otherwise run via the asynchronous command server, don't race with
+              the keyboard.
+            - But this isn't important for the Cursorless everywhere user, for
+              whom VS Code only exists to run in the background and handle
+              cursorless commands.
+        - That should unblock @rntz for now.
+
+3. Right now hats in VS Code ("editor decorations") effectively get rebased
+   whenever the user makes changes to the document -- VS Code does that for
+   free.
+    - For example, if you press enter on the line above a series of hats, the
+      hats move down alongside the line instantly -- there's no jank.
+    - This doesn't happen with the JetBrains prototype, so you instead have hats
```

Comment:
oh interesting.  I guess that's a side effect of the way you render them in JetBrains?  Assuming you use global positions?  In VScode they are just css `:before` pseudotags so they move along with their span (ie the text underneath the hat)

Reply:
Right. We simply have line/column ranges as numbers disassociated with content, so when the content changes and they don't, they look strange. I think VS Code decoration ranges connect to them to the content implicitly and moves them for you automatically.

I'm going to revisit this once I get the temporary file working to see how bad it is if we just wait for new hats to be generated, but I suspect it would probably be worth writing some code to mimic the same behavior, just to reduce the jank. I think we just need to watch for insertions or deletions, and move the hats that proceed/follow those ranges by the same amount.